L3DAS22 enhancement  Recipe

neillu23 opened the pull request.

Emrys365 commented on egs2/l3das22/enh1/local/l3das22_multi_channel.sh:
```diff
@@ -0,0 +1,83 @@
+set -o pipefail
+
+# configs
+nj=16
```
If this is a parsable argument, it is better to add a `help_message` string like in [egs2/chime4/enh1/local/data.sh#L13-L30](https://github.com/espnet/espnet/blob/master/egs2/chime4/enh1/local/data.sh#L13-L30).

neillu23 replied:
Thanks! Added in 
https://github.com/neillu23/espnet/blob/l3das22/egs2/l3das22/enh1/local/l3das22_multi_channel.sh

Emrys365 reviewed (commented):
Thanks! I just made some minor comments.

Emrys365 commented:
Could you also add your recipe to [egs2/README.md](https://github.com/espnet/espnet/blob/master/egs2/README.md)?

sw005320 commented on egs2/l3das22/enh1/local/normalize_transcript.pl:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env perl
```
is it needed?

sw005320 reviewed (commented):
This recipe should have a special evaluation (scoring) script prepared by an organizer (or by making the exact same scoring scheme by ourselves).

neillu23 commented:
Thanks for the review and comments! I have addressed the comments and made two major changes to the PR:
1. Add the testing data from the L3DAS challenge.
2. Add metric.sh for calculating the challenge score.

sw005320 commented on egs2/l3das22/enh1/local/metric.py:
```diff
@@ -0,0 +1,139 @@
+import jiwer
+import numpy as np
+from pystoi import stoi
+import transformers
```
Is it OK?
I think we need to install something.

neillu23 replied:
Yes, we need to install jiwer (WER), pystoi (STOI), and transformers (ASR).
I can add an installation part to the metric.sh.

sw005320 commented on egs2/l3das22/enh1/local/metric.sh:
```diff
@@ -0,0 +1,20 @@
+fi
+
+if [ $# -gt 1 ]; then
+	predicted_path=$1
+	target_path=$2
+  python local/metrics.py --predicted_path ${predicted_path} --target_path ${target_path}
+else
+	python local/metrics.py
```
The indent looks wrong.

neillu23 replied:
Thanks for pointing it out. This didn't happen in my code viewer.

Emrys365 commented on egs2/l3das22/enh1/local/metric.sh:
```diff
@@ -0,0 +1,24 @@
+. parse_options.sh || exit 1;
+
+pip show -f jiwer >/dev/null || pip install jiwer
+pip show -f pystoi >/dev/null || pip install pystoi
```
I think `pystoi` is ensured to be installed in setup.py

neillu23 replied:
Thanks! Let me remove it.

ftshijt commented on egs2/l3das22/enh1/local/data.sh:
```diff
@@ -0,0 +1,63 @@
+}
+
+
+help_message=$(cat << EOF
+Usage: $0 
+  optional argument:
+    None
+EOF
+)
```
Maybe just a warning message for calling with other arguments? It seems not a large problem

neillu23 replied:
Added.

ftshijt commented on egs2/l3das22/enh1/local/data.sh:
```diff
@@ -0,0 +1,63 @@
+if [ ! -e "${L3DAS22}" ] ; then
+    log "
+    Please fill the value of 'L3DAS22' in db.sh
+    The 'L3DAS22' (https://www.kaggle.com/l3dasteam/l3das22) directory 
+    should at least contain the task1 dataset:
```
In that case, does it possible to directly download the data from kaggle API?

neillu23 replied:
Added downloading the data from kaggle.

ftshijt commented on egs2/l3das22/enh1/local/l3das22_data_prep.sh:
```diff
@@ -0,0 +1,102 @@
+
+L3DAS22=$1
+
+tmpdir=data/temp
```
`mktemp` would be better in the case, and please also use trap to delete the corresponding folder after the preparation.

neillu23 replied:
Since the files under the temp directory are used in the following recipe (local/l3das22_multi_channel.sh), do you think we could keep using this directory for passing the files? (or we could move mktemp to local/data.sh)

ftshijt commented on egs2/l3das22/enh1/local/l3das22_multi_channel.sh:
```diff
@@ -0,0 +1,93 @@
+
+echo "Building multi-channel data"
+
+tmpdir=data/temp
```
same for here, `mktemp` seems to fit the context better

ftshijt commented on egs2/l3das22/enh1/local/metric.sh:
```diff
@@ -0,0 +1,23 @@
+[ -f ./path.sh ] && . ./path.sh
+. parse_options.sh || exit 1;
+
+pip show -f jiwer >/dev/null || pip install jiwer
+pip show -f transformers >/dev/null || pip install transformers
```
We may add a note that these evaluation metrics are from the official requirements?

neillu23 replied:
Added.

ftshijt commented on egs2/l3das22/enh1/pyscripts:
```diff
@@ -0,0 +1 @@
+../../TEMPLATE/asr1/pyscripts
```
though may not be different, by use `enh1/pyscripts` would be better (just in case future differences)

neillu23 replied:
Changed.

ftshijt reviewed (commented):
Looks very cool! some minor suggestion

neillu23 commented:
Thanks for the comments! I have added the downloading data part in the recipe!

kamo-naoyuki commented on egs2/l3das22/enh1/run.sh:
```diff
@@ -0,0 +1,26 @@
+    --spk_num 1 \
+    --enh_config conf/tuning/train_enh_beamformer_mvdr.yaml \
+    --use_dereverb_ref false \
+    --use_noise_ref true \
```
I have a question.

In my run, noise.scp was not generated. How can I get it?

Emrys365 replied:
I don't think we have access to the noise data. So it should be set to false here.

kamo-naoyuki replied:
Okay, thanks!

neillu23 replied:
Yes, I have changed use_noise_ref to false. Thanks for the help!

sw005320 commented:
@neillu23, some comments still exist, including the results.

neillu23 commented:
> @neillu23, some comments still exist, including the results.

Thanks for reminding me! I haven't changed the data/temp to `mktemp` in local/l3das22_data_prep.sh because the files under the temp directory are used in the following script (local/l3das22_multi_channel.sh). Do you think we should pass the temp directory variable across the scripts?
I am training a fasnet and I will update the results soon.

Emrys365 commented on egs2/l3das22/enh1/local/metric.py:
```diff
@@ -0,0 +1,136 @@
+import sys
+import warnings
+
+import jiwer
+import numpy as np
+import soundfile as sf
+import torch
+import torchaudio
+import transformers
+from pystoi import stoi
+from tqdm import tqdm
+from transformers import Wav2Vec2ForMaskedLM, Wav2Vec2Tokenizer
+
```
Could you apply `isort` to sort the imports?

Emrys365 reviewed (approved):
LGTM in general. I think we can merge it after existing comments are addressed.

Emrys365 commented:
LGTM! I will merge it. Thanks a lot @neillu23
